simulator/Coverage.kt

```diff
@@ -0,0 +1,56 @@
+package venusbackend.simulator
+
+import venusbackend.linker.ProgramDebugInfo
+import venusbackend.riscv.MachineCode
+import venusbackend.toHex
+import java.io.File
+
+class Coverage(private val sim: Simulator, private val outputFile: String) : IsSimulatorPlugin {
+
+    private val pcCount = mutableMapOf<Number, Int>()
+
+    init {
+        assert(outputFile.isNotBlank()) {"Blank output file for coverage: $outputFile"}
+    }
+
+    override fun onStep(inst: MachineCode, prevPC: Number) {
+        val newCount = pcCount.getOrDefault(prevPC, 0) + 1
+        pcCount[prevPC] = newCount
+    }
+
+    /** saves the coverage to the coverage file */
+    fun finish() {
+        // first we find all locations
+        val allLocations = sim.linkedProgram.dbg.map { d -> getLocation(d) }
+
+        // map location to count + pc
+        val locationToCount = pcCount.map { (pc, count) -> getLocation(pc) to count }.toMap()
+
+        val lines = allLocations.withIndex().map { (idx, loc) ->
+            val pc = instructionIndexToPc(idx)
+            val count = locationToCount.getOrDefault(loc, 0)
+            "${toHex(pc)} $loc $count"
+        }
+
+        File(outputFile).bufferedWriter().use { out ->
```

Comment:
you should be using the Virtual (Venus) File System instead as that is the abstraction for file operations for js and jvm compatability

Reply:
I am moving the code to write the output out of the backend, finish is going to return a `List<String>`